### PR TITLE
NFC: Clean up WhereClauseOwner

### PR DESCRIPTION
**Pull request by AnthonyLatsis**
@slavapestov 

Midquel for https://github.com/apple/swift/pull/23489

Apart from the clean up, this will allow us to properly pass a `GenericContext` rather than just a Decl. As far as I see directly passing a `TrailingWhereClause` is not an option, because `WhereClauseOwner` *needs* a real `DeclContext`.

### CODE REVIEW

**Comment by AnthonyLatsis** [comment]
Ugh, formatting conflicts

File: lib/Sema/TypeCheckAccess.cpp [hunk 1]
```diff
@@ -379,9 +379,9 @@ void AccessControlCheckerBase::checkGenericParamAccess(
     auto diagID = diag::generic_param_usable_from_inline;
     if (downgradeToWarning == DowngradeToWarning::Yes)
       diagID = diag::generic_param_usable_from_inline_warn;
-    auto diag = TC.diagnose(owner,
+    auto diag = TC.diagnose(ownerDecl,
```
Comment (CodaFi) [inline lib/Sema/TypeCheckAccess.cpp]: `ownerDecl->diagnose` please.

Comment (CodaFi) [reply lib/Sema/TypeCheckAccess.cpp]: Actually, this will fall out of the merge conflict resolution...

Comment (AnthonyLatsis) [reply lib/Sema/TypeCheckAccess.cpp]: Indeed, I've spotted it now.

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by AnthonyLatsis** [comment]
🤔

**Comment by AnthonyLatsis** [comment]
Found the culprit, let's go again.

**Comment by theblixguy** [comment]
@swift-ci please smoke test

**Review by DougGregor — APPROVED** [review]

**Comment by CodaFi** [comment]
⛵️

**Pull request merged**